In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

df = pd.read_csv("../../completed/df_stats.csv", index_col=0)

## Parte 2 - Modellazione e Predizione

Nella prima parte abbiamo conosciuto il dataset, ci abbiamo lavorato modificandolo in modo da essere più adatto per questa seconda parte.

Adesso ci occuperemo della **modellazione**, gli obiettivi sono:
- **Understanding** (Approccio Statistico): utilizzo di modelli di regressione per quantificare le relazioni tra le variabili, testare ipotesi e comprendere l'impatto dei fattori in gioco;
- **Prediction** (Approccio Machine Learning): Costruire sistemi automatici in grado di stimare valori basandosi su dati nuovi mai visti oppure classificarli con l'obiettivo di generare un valore pratico;

### 2.1 Analisi Statistica e Regressione (Understanding)

Iniziamo con l'approccio statistico, l'attenzione verrà posta sulla significatività statistica e sull'interpretazione delle varie variabili (coefficienti).

Non stiamo ancora cercando di fare la "miglior predizione possibile", ma di capire come le variabili si influenzano a vicenda.

**Selezione delle variabili**

Come abbiamo detto in precedenza, la nostra variabile target è Price (il prezzo di vendita delle auto), quindi cercheremo di capire a fondo l'impatto che hanno le varie variabili sul nostro target.

In [3]:
y_inference = df['Price']

vars_da_escludere = ['Price'] 
X_inference = df.drop(columns=vars_da_escludere)

X_inference = sm.add_constant(X_inference)

modello = sm.OLS(y_inference, X_inference).fit()

print(modello.summary())

                            OLS Regression Results                            
Dep. Variable:                  Price   R-squared:                       0.932
Model:                            OLS   Adj. R-squared:                  0.932
Method:                 Least Squares   F-statistic:                     2171.
Date:                Fri, 02 Jan 2026   Prob (F-statistic):               0.00
Time:                        18:34:10   Log-Likelihood:                 1362.4
No. Observations:                5710   AIC:                            -2651.
Df Residuals:                    5673   BIC:                            -2405.
Df Model:                          36                                         
Covariance Type:            nonrobust                                         
                          coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------
const                  -0.8504    

Abbiamo applicato un modello di Regressione ai nostri dati, abbiamo utilizzat il metodo dell'Ordinary Least Squares (OLS), analizziamo cosa abbiamo ottenuto.

**Goodness of Fit**:
- R-squared (0.932): il modello spiega il 93.2% della varianza dei prezzi, indica un adattamento ottimo dei dati.
- Prob. F-statistic (0.00): il modello è statisticamente significativo.
- Adj. R-squared (0.932): valore uguale ad R-squared, significa che non abbiamo "gonfiato" il modello con variabili inutili, tutte portano almeno un po' di informazione.

**Multicollinearità**

Inoltre vengono restituite due note, esse riguardano un pericolo di possibile multicollinearità, in cui più variabili potrebbero offrire la stessa informazione rendendo i risultati inconsistenti.